core, editoast, python: stop train at end of block instead of OP

louisgreiner opened the pull request:
Closes https://github.com/osrd-project/osrd-confidential/issues/892
Closes https://github.com/osrd-project/osrd-confidential/issues/830

Analysis of the solution, compared to the default behavior can be seen [here](https://github.com/osrd-project/osrd-dags/pull/655)

## Description

Note : the following screenshots are for review visualization only, nothing will be output in the front-end at the end

The trains were previously stopping exactly on the Operational Point Part position, which is often in the middle of the train platform. The position of the train is considered at the head of the train. This is an issue, because in the real life, a train (meaning its head) stops at the end of the platform (at a specific stop panel position), allowing the rest of the train not to overflow of the passed tracks, avoiding unnecessary track occupation. Also, this doesn't take account of the train direction.

The train stops are now shifted the least possible to the next signal position, with a minimal value that would put the tail of the train on the operational point.

This implementation makes the pathfinding algorithm move the stops of these trains closer to the end of the current block, to try to avoid that the tail of the train is on another track, and also never further than the next block-delimiting signal.
⚠️ the first and last stops are not moved ⚠️

- The feature is ✅ hard-enabled for STDCM (for the simulated train), and there is no deactivation possible.
- The feature is ✖️ hard-disabled in front for Operational Studies (trains created from this module cannot enable the feature)
  - however the feature is 🔲 activable in back (disabled by default), so imported trains that have the feature enabled (such as BASIC timetables) can be looked at in Operational Studies, or used in STDCM search environment.
- The feature may (☑️ will) be enabled for all the trains simulated in the STDCM search environment depending on what was parametrized when importing data.

## Examples

#### With a `TGV2N2`, 400m train
![image](https://github.com/user-attachments/assets/f17049a0-03ad-4db8-a88d-ce0277fa7af8)

#### With a `83500`, 109m train
![image](https://github.com/user-attachments/assets/29c3a370-f132-41dd-a773-344406b1a7a9)

#### Also working on STDCM, with a 200m train
![image](https://github.com/user-attachments/assets/60610f4f-ac9c-460c-8691-0b443d5a3cd6)

louisgreiner commented on editoast/editoast_schemas/src/train_schedule/train_schedule_options.rs:
```diff
@@ -14,8 +14,22 @@ pub struct TrainScheduleOptions {
     #[serde(default = "default_use_electrical_profiles")]
     use_electrical_profiles: bool,
+
+    #[derivative(Default(value = "true"))]
```
```suggestion
    #[derivative(Default(value = "false"))]
```

We keep this flag to false for review purpose, but before merging, we need to set it to `false`

louisgreiner commented on editoast/editoast_schemas/src/train_schedule/train_schedule_options.rs:
```diff
@@ -14,8 +14,22 @@ pub struct TrainScheduleOptions {
     true
 }
+
+fn default_stop_at_next_signal() -> bool {
+    true
+}
```
```suggestion
```

We keep this flag to `true` for review purpose, but before merging, we need to set it to `false`

louisgreiner replied:
```suggestion
```

louisgreiner commented on front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:
```diff
@@ -149,6 +149,7 @@ const useSetupItineraryForTrainUpdate = (trainIdToEdit: number) => {
           rolling_stock_supported_signaling_systems: rollingStock.supported_signaling_systems,
           rolling_stock_maximum_speed: rollingStock.max_speed,
           rolling_stock_length: rollingStock.length,
+          stop_at_next_signal: true, // TODO: change to false and set to true elsewhere if needed
```
```suggestion
          stop_at_next_signal: false,
```

We keep this flag to `true` for review purpose, but before merging, we need to set it to `false`

louisgreiner commented on front/src/modules/pathfinding/utils.ts:
```diff
@@ -79,6 +79,7 @@ export const getPathfindingQuery = ({
         rolling_stock_supported_signaling_systems: rollingStock.supported_signaling_systems,
         rolling_stock_maximum_speed: rollingStock.max_speed,
         rolling_stock_length: rollingStock.length,
+        stop_at_next_signal: true, // TODO: change to false and set to true elsewhere if needed
```
```suggestion
        stop_at_next_signal: false,
```

We keep this flag to `true` for review purpose, but before merging, we need to set it to `false`

louisgreiner replied:
```suggestion
        stop_at_next_signal: false,
```

louisgreiner commented on editoast/editoast_schemas/src/train_schedule/train_schedule_options.rs:
```diff
@@ -14,8 +14,22 @@ pub struct TrainScheduleOptions {
     #[serde(default = "default_use_electrical_profiles")]
     use_electrical_profiles: bool,
+
+    #[derivative(Default(value = "true"))]
+    #[serde(default = "default_stop_at_next_signal")]
```
```suggestion
    #[derivative(Default(value = "false"))]
```

woshilapin replied:
Hmm, I'm curious, but why not keeping `true` as a default value? Seems (naively maybe) like a good default, isn't it?

flomonster replied:
The current behavior is `false`. If we set the new default behavior to `true` we need to add a migration to this PR. I do prefer to keep the behavior as `false`.

louisgreiner commented on front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:
```diff
@@ -149,6 +149,7 @@ const useSetupItineraryForTrainUpdate = (trainIdToEdit: number) => {
           rolling_stock_supported_signaling_systems: rollingStock.supported_signaling_systems,
           rolling_stock_maximum_speed: rollingStock.max_speed,
           rolling_stock_length: rollingStock.length,
+          stop_at_next_signal: true, // TODO: change to false and set to true elsewhere if needed
```
```suggestion
          stop_at_next_signal: false,
```

clarani replied:
Same here

louisgreiner reviewed (commented):
We keep this flag to `true` for review purpose, but before merging, we need to set it to `false`

louisgreiner commented:
Questions : 
- Currently it moves all the waypoints of the train schedules instead of only the stops ; is it a real problem ?

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +384,37 @@ private fun getBlockOffset(
+
+    // some blocks are < 1m long (even 0m), we can't get further in the block
+    val maxHeadOffset =
+        if (blockOffset.distance < 1.meters) {
+            blockOffset.distance
```
That `if` feels wrong, if the train starts at the beginning of the block, then we don't use the next signal offset?

louisgreiner replied:
You are right indeed, it sucks 😄 

Also, I couldn't manage otherwise to deal with that later in the code, if the train stops **exactly** on one of a border of the block (either `start` or `end` if block is `[start ; end]`), I then don't know how it will react.
That's why I introduced a magic value (1m) to make sure this will not happen.

Do you have any better solution ? Or insight

eckter replied:
I don't think we should really handle it in any specific way.

When the point is exactly at the end of a block, there's not much we can do anyway to move it further. It's supposed to be near a signal or buffer stop anyway. 

When it's at the beginning of a block, we can handle it normally by moving it forward using the same rules as usual.

eckter reviewed (commented):
That's the right approach, good job :+1:

louisgreiner commented:
Question : is the type of signalisation (here `TVM` instead of `BAL`) could affect the behaviour of the code ? 

![image](https://github.com/user-attachments/assets/ee148e0e-76d5-4c65-892e-16e1f87e2785)

Because here, the waypoint is moved AFTER the signal, which is not correct (but I only see that behaviour on `TVM` signals)

Could it be also related to the current bug that breaks the projection of intermediary waypoints on the map ?

Edit: looks like I was right, but difficult to be sure (the position here is now correct for the end pathstep)

![image](https://github.com/user-attachments/assets/27988c5b-24ec-4690-bcf0-5356fe116089)

woshilapin reviewed (approved):
Looks good. I'm just wondering about making it a default option?

flomonster commented on editoast/src/views/path/pathfinding.rs:
```diff
@@ -472,7 +476,8 @@ pub mod tests {
                 "rolling_stock_maximum_speed":22.00,
-                "rolling_stock_length":26.00
+                "rolling_stock_length":26.00,
+                "stop_at_next_signal":false
```
We can omit this and use the default value.

Caracol3 reviewed (approved):
LGTM

clarani commented on front/src/modules/pathfinding/utils.ts:
```diff
@@ -84,6 +84,7 @@ export const getPathfindingQuery = ({
         rolling_stock_supported_signaling_systems: rollingStock.supported_signaling_systems,
         rolling_stock_maximum_speed: rollingStock.max_speed,
         rolling_stock_length: rollingStock.length,
+        stop_at_next_signal: true, // TODO: change to false and set to true elsewhere if needed
```
Hmm I think it should be false by default ?

louisgreiner replied:
Indeed, it wrote above a comment about this : I kept it `true` to make it easier to test, but I will update these values at `false` as soon as the tests are done from the reviewers' side

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,41 @@ private fun getBlockOffset(
+    return PathfindingEdgeLocationId(waypointBlock.edge, nextSignalOffset)
+}
+
+private fun getNextSignalOffset(
+    blockId: BlockId,
+    blockOffset: Offset<Block>,
+    infra: FullInfra,
+    rollingStockLength: Double
+): Offset<Block> {
```
This needs some documentation, it's quite unclear why a function called `getNextSignalOffset()` cares about the `rollingStockLength`

louisgreiner replied:
You're totally right ; I moved and renamed some of this stuff, and also re-wrote the documentation attached. Is that ok that way ?

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,41 @@ private fun getBlockOffset(
 }
+
+/**
+ * Returns the block of the signal next to a waypoint block.
```
This doesn't return a Block

louisgreiner replied:
I rephrased, in 
```kt
 * @return stop position on the signal next to the waypoint block.
```
But I also could have had been more concise : 

```kt
 * @return the stop position
```

What do you think ?

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,41 @@ private fun getBlockOffset(
+    val nextSignalPosition = signalsPositions.firstOrNull { it.distance >= blockOffset.distance }
+
+    val minTailOffset = blockOffset.distance + rollingStockLength.meters
+    val maxHeadOffset = if (blockOffset.distance < 0.01.meters) {
```
We should probably use `10.millimeters` instead of a floating point approximation. (It should also be a constant instead of a magic number)

Erashin replied:
It is unclear whether this is still necessary or not in the overall PR. I'm down for a check/quick explanation :)

louisgreiner replied:
It has been removed, since a heavenly fix came meanwhile

Khoyo commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,41 @@ private fun getBlockOffset(
+    val maxHeadOffset = if (blockOffset.distance < 0.01.meters) {
+        blockOffset.distance
+    } else {
+        (nextSignalPosition?.distance ?: blockLength) - 0.01.meters
```
Ditto

flomonster commented on editoast/editoast_schemas/src/train_schedule/train_schedule_options.rs:
```diff
@@ -14,8 +14,17 @@ pub struct TrainScheduleOptions {
     #[serde(default = "default_use_electrical_profiles")]
     use_electrical_profiles: bool,
+
+    #[derivative(Default(value = "false"))]
```
1. Default `bool` is already `false`. Check the doc [here](https://doc.rust-lang.org/std/primitive.bool.html#impl-Default-for-bool).
2. We should indicate to serde that the value can be omitted. 

```suggestion
    #[serde(default)]
```

louisgreiner replied:
thanks for the explanation :-)

bougue-pe commented on editoast/src/core/pathfinding.rs:
```diff
@@ -42,6 +42,8 @@ pub struct PathfindingRequest {
     pub rolling_stock_maximum_speed: f64,
     /// Rolling stock length in meters:
     pub rolling_stock_length: f64,
+    /// If the train should stop on the next signal instead of on the operational point
```
nit: This could match the comment from editoast's API (which may be rephrased as mentioned in another comment)

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,38 @@ private fun getBlockOffset(
+ * Returns the stop position of a train on the signal next to a waypoint block, considering its
+ * rolling stock length.
+ *
+ * @param waypointBlock waypoint block.
+ * @param rollingStockLength length of the rolling stock.
+ * @param infra full infra.
+ * @return stop position on the signal next to the waypoint block.
```
nit: Those can be removed as, to my knowledge, it's not a target to document all inputs/outputs in core, and those don't add information to the existing code or comment above.

louisgreiner replied:
removed then, I thought we add documentation to public functions

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,38 @@ private fun getBlockOffset(
+
+/**
+ * Returns the stop position of a train on the signal next to a waypoint block, considering its
+ * rolling stock length.
```
As:
* waypoint block is not so obvious name
* it does more than just get the position of the next signal
```suggestion
 * Given a waypoint block (initial position of a train stop), returns the new stop position of a train: next signal, with the constraint of staying in the same block and keeping the tail of the train on the initial position.
 * Note that nodes of the infra are not considered.
```

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,38 @@ private fun getBlockOffset(
+    infra: FullInfra
+): PathfindingEdgeLocationId<Block> {
+    // To ensure that the tail of the rolling stock is not further back than the operational point
+    val minTailOffset = waypointBlock.offset.distance + rollingStockLength.meters
```
It's more of a max limit, no? (applying the min to the 2 max values at the end)
```suggestion
    val maxTailOffset = waypointBlock.offset.distance + rollingStockLength.meters
```

louisgreiner replied:
Totally right !! 💯

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,38 @@ private fun getBlockOffset(
+    rollingStockLength: Double,
+    infra: FullInfra
+): PathfindingEdgeLocationId<Block> {
+    // To ensure that the tail of the rolling stock is not further back than the operational point
```
nit:
```suggestion
    // To ensure that the tail of the rolling stock is not further back than the operational point (waypointBlock)
```

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlocksEndpointV2.kt:
```diff
@@ -369,3 +385,38 @@ private fun getBlockOffset(
+    // To ensure that the tail of the rolling stock is not further back than the operational point
+    val minTailOffset = waypointBlock.offset.distance + rollingStockLength.meters
+
+    // If there is no signal after the block, we consider the end of the block as the stop position
```
```suggestion
    // If there is no signal in the block after the operational point (waypointBlock), we consider the end of the block as the new stop position
```

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+        val waypointsIntermediate = listOf(waypointIntermediate)
+        val waypointsEnd = listOf(waypointEnd)
+        val waypoints = listOf(waypointsStart, waypointsIntermediate, waypointsEnd)
+        val reverseWaypoints = listOf(waypointsEnd, waypointsIntermediate, waypointsStart)
```
I'd use `waypoints.reversed()` directly, instead.
(then maybe remove the 3 intermediate lists `waypointsStart`, etc. is a valid option)

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+        val reverseWaypoints = listOf(waypointsEnd, waypointsIntermediate, waypointsStart)
+
+        val nonStopAtNextSignalRequestBody =
+            pathfindingRequestAdapter.toJson(
+                PathfindingBlockRequest(
+                    rollingStockLoadingGauge = RJSLoadingGaugeType.G1,
+                    rollingStockIsThermal = true,
+                    rollingStockSupportedElectrifications = listOf(),
+                    rollingStockSupportedSignalingSystems = listOf("BAL"),
+                    rollingStockMaximumSpeed = 320.0,
+                    rollingStockLength = 200.0,
+                    timeout = null,
+                    infra = "small_infra/infra.json",
+                    expectedVersion = "1",
+                    pathItems = waypoints,
+                    stopAtNextSignal = false,
+                )
+            )
```
nit: This could be mutualized in a free fn.

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+            (pathfindingResponseAdapter.fromJson(reversedLongTrainResponse)
+                as? PathfindingBlockSuccess)!!
+
+        assertEquals(3, nonStopAtNextSignalParsed.pathItemPositions.size)
+        assertEquals(0.meters, nonStopAtNextSignalParsed.pathItemPositions[0].distance)
+        assertEquals(12050.meters, nonStopAtNextSignalParsed.pathItemPositions[1].distance)
+        assertEquals(26500.meters, nonStopAtNextSignalParsed.pathItemPositions[2].distance)
```
nit: I would use this one also for the reversed case: it tests/explains the control-case.

louisgreiner replied:
Right, added

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+        assertEquals(3, longTrainParsed.pathItemPositions.size)
+        assertEquals(0.meters, longTrainParsed.pathItemPositions[0].distance)
+        assertEquals(12250.meters, longTrainParsed.pathItemPositions[1].distance)
+        assertEquals(26517.5.meters, longTrainParsed.pathItemPositions[2].distance)
```
* Is this the only case tested on the signal?
* Should we also test the block-end?

louisgreiner replied:
Using small infra, it's hard to test more hazardous cases :/

bougue-pe replied:
As discussed:
* this does test the signal case indeed.
* we should add a test on the "end-of-block" case, and the easiest is to use a very long (600m) train on the reverse case and check that it stops after 500m (buffer-stop of TA1)

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+
+        assertEquals(3, shortTrainParsed.pathItemPositions.size)
+        assertEquals(0.meters, shortTrainParsed.pathItemPositions[0].distance)
+        assertEquals(12060.meters, shortTrainParsed.pathItemPositions[1].distance)
```
nit for explicitness:
How about defining before (ditto for other values):
```kt
val shortTrainLength = 10.meters
val initialIntermediateStop = 12050.meters
```
then use `initialIntermediateStop + shortTrainLength`

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+                )
+            )
+
+        val nonStopAtNextSignalRawResponse =
+            PathfindingBlocksEndpointV2(infraManager)
+                .act(RqFake("POST", "/v2/pathfinding/blocks", nonStopAtNextSignalRequestBody))
+        val nonStopAtNextSignalResponse =
+            TakesUtils.readBodyResponse(nonStopAtNextSignalRawResponse)
+        val nonStopAtNextSignalParsed =
+            (pathfindingResponseAdapter.fromJson(nonStopAtNextSignalResponse)
+                as? PathfindingBlockSuccess)!!
```
nit: Mutualizing this would be nice too.

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingV2Test.kt:
```diff
@@ -239,4 +244,165 @@ class PathfindingV2Test : ApiTest() {
+        assertEquals(12050.meters, nonStopAtNextSignalParsed.pathItemPositions[1].distance)
+        assertEquals(26500.meters, nonStopAtNextSignalParsed.pathItemPositions[2].distance)
+
+        assertEquals(3, shortTrainParsed.pathItemPositions.size)
```
nit: Also, I would rather have each case separated (one test fn for each, or at least group body+request+assert)

louisgreiner replied:
Done, is it conform to what you expected ?

bougue-pe replied:
Yes, thank you very much!

bougue-pe commented on editoast/src/views/path/pathfinding.rs:
```diff
@@ -75,6 +75,8 @@ struct PathfindingInput {
     /// Rolling stock length
     #[schema(value_type = f64)]
     rolling_stock_length: OrderedFloat<f64>,
+    /// Stops the train at next signal instead of on path item
```
As it does more than just this (see suggestion of code comment in core), it would be nice to elaborate more in the doc of the API.

bougue-pe commented on python/osrd_schemas/osrd_schemas/train_schedule.py:
```diff
@@ -170,12 +170,19 @@ class PowerRestrictionRanges(RootModel):
     )
+    stop_at_next_signal: bool = Field(
+        default=False,
+        description="If true, the train will stop at the next signal instead of at the operational point",
```
Same here, it would be nice to have the same description than editoast's API (and more explicit on what's done).

bougue-pe reviewed (commented):
🎉 Nice job (especially with the side-bug messing around)

Just comments on the form, the content is good to me.
Feel free to reject the nitpicks, they're not mandatory IMO and sometimes just a matter of taste.

bougue-pe reviewed (approved):
Thanks!

bougue-pe commented on front/tests/assets/trainSchedule/train_schedules.json:
```diff
@@ -92,7 +92,7 @@
     "speed_limit_tag": "HLP",
     "power_restrictions": [],
-    "options": { "use_electrical_profiles": false }
+    "options": { "use_electrical_profiles": true, "stop_at_next_signal": false }
```
From previous value, it looks like a bad copy-paste:
```suggestion
    "options": { "use_electrical_profiles": false, "stop_at_next_signal": false }
```

louisgreiner replied:
oops, thanks for the spot 🎯

bougue-pe commented on front/tests/assets/trainSchedule/train_schedules.json:
```diff
@@ -264,7 +264,7 @@
     "speed_limit_tag": null,
     "power_restrictions": [],
-    "options": { "use_electrical_profiles": false }
+    "options": { "use_electrical_profiles": true, "stop_at_next_signal": false }
```
```suggestion
    "options": { "use_electrical_profiles": false, "stop_at_next_signal": false }
```

bougue-pe commented on front/tests/assets/trainSchedule/train_schedules.json:
```diff
@@ -493,7 +493,7 @@
     "speed_limit_tag": "HLP",
     "power_restrictions": [],
-    "options": { "use_electrical_profiles": false }
+    "options": { "use_electrical_profiles": true, "stop_at_next_signal": false }
```
```suggestion
    "options": { "use_electrical_profiles": false, "stop_at_next_signal": false }
```

bougue-pe commented on front/tests/assets/trainSchedule/train_schedules.json:
```diff
@@ -672,6 +672,6 @@
     "speed_limit_tag": "MA100",
     "power_restrictions": [],
-    "options": { "use_electrical_profiles": false }
+    "options": { "use_electrical_profiles": true, "stop_at_next_signal": false }
```
```suggestion
    "options": { "use_electrical_profiles": false, "stop_at_next_signal": false }
```

louisgreiner commented:
## Feature not ready for STDCM simulation

Last commit `core: add feature flag to disable feature for stdcm` is meant to disable the feature for the simulated trains of STDCM. 

The issue is that, after having moved the waypoints, we can't match the new track positions with actual operational points. Then, the output tables can't match the stops information of the newly moved waypoints with actual operational points.

Example what we should get versus what we get with the feature:
![Capture d'écran 2025-02-07 110951](https://github.com/user-attachments/assets/f2e79184-ad85-4d80-ad5f-7b159ac56ebd)
![Capture d'écran 2025-02-07 111552](https://github.com/user-attachments/assets/ba0d3374-b16a-48d3-9141-96d2e51f2ed4)

However, it is still not an ideal trade-off, because with the feature disabled for STDCM only, if we want to debug the STDCM timetables, we still face this issue on the Operational Studies module.

Example, the stop information has disapeared from the output table, because, and we see it, the waypoint has moved:
![Capture d'écran 2025-02-07 165731](https://github.com/user-attachments/assets/790febfc-934c-4bf6-990a-cbf9b2c121e4)
![Capture d'écran 2025-02-07 165744](https://github.com/user-attachments/assets/0d537511-6e45-4772-b10f-cedbd7404dea)
![Capture d'écran 2025-02-07 165851](https://github.com/user-attachments/assets/3fdbc521-34e3-4691-89ca-abd8e6500176)

## Feature still interesting for filling the STDCM timetable

The feature remains enabled for simulating the trains of the timetable STDCM stands on, thus potentially gaining still a good part of the capacity the feature should provide. But as mentionned above, it breaks the debug functionnality for STDCM timetable for the stops.

With the [current state of code](https://github.com/OpenRailAssociation/osrd/blob/dev/core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPostProcessing.kt#L104-L106), it is difficult not to break anything to make the new positions of waypoints match with actual waypoints.

flomonster reviewed (approved):
LGTM for the editoast part, we should add a migration to drop this field (up.sql should be empty).

Synar reviewed (approved):
E2e-tests : LGTM!

Commit:
core, editoast, front: add stops_at_end_of_block option to TrainSchedule

This commit uses `stops_at_end_of_block` flag to deal with the issue that the trains were previously stopping exactly on the Operational Point Part position, which is often in the middle of the train platform. The position of the train is considered at the head of the train. This is an issue, because in the real life, a train (meaning its head) stops at the end of the platform (at a specific stop panel position), allowing the rest of the train not to overflow of the passed tracks, avoiding unnecessary track occupation. Also, this doesn't take account of the train direction.

This implementation makes the pathfinding algorithm move the stops of these trains closer to the end of the current block, to try to avoid that the tail of the train is on another track, and also never further than the next block-delimiting signal.

- Operational Studies:
	- Nothing changed except option propagation: `stops_at_end_of_block = false` (default).
- STDCM:
	- The trains that compose the timetable (search environment) are all simulated with `stops_at_end_of_block = true`.
	- The simulated train is simulated also with this flag `stops_at_end_of_block = true`.

Note: this should be reverted when we have the stop panels.

Signed-off-by: Louis Greiner <greiner.louis@gmail.com>
Signed-off-by: Angelina Kuntz <angelina.kuntz@epita.fr>

louisgreiner commented:
Thanks to @Synar [PR](https://github.com/OpenRailAssociation/osrd/pull/12474) that fixes the issue on the operational studies output table (the moved stops were not displayed in the output table, when debugging an STDCM timetable), this PR will soon come to its end 🤞 

We need to wait that the above PR is merged, then final review can be done. I also fixed some tests to make sure that only the intermediate stops are moved close to the following signal (and not the start nor the last stop).

louisgreiner commented:
Dear reviewers, this PR is finally working 🎉 Because the PR is old, it might be necessary to review it once again.

Quick recap:
- this PR enables to move the stops of a pathfinding closer to the signal next to the requested stop
  - ⚠️ **the first and last stops are not moved** ⚠️
- the feature is enabled for STDCM (for all the trains simulated in the stdcm search environment AND for the simulated train)
- the feature is disabled for Operational Studies (trains created from this module cannot enable the feature, however imported trains that have the feature enabled (such as BASIC timetables) can be looked at in this module)
- [osrd-data associated MR](https://gitlab-repo-res.apps.eul.sncf.fr/dsir/groupedxs-dsir/04735/osrd-data/-/merge_requests/8)

To test this PR:
- either import this timetable in the Operational Studies module (you can first look at the stops on the map that they are indeed moved, train `"870004"` that stops at `"Gragnague"`, [timetable_chenille.json](https://github.com/user-attachments/files/21599681/timetable_chenille.json)), use this scenario as STDCM search environment (debug mode) and simulate trains in STDCM
- either locally import a BASIC timetable and run STDCM on it (also take a look at this timetable in OP module)

Questions: 
- should I do something about the [API changes automatic comment](https://github.com/OpenRailAssociation/osrd/pull/10200#issuecomment-3153925166)?
- ETCS simulations are missing in this PR, should it be part of it?

Khoyo commented:
> should I do something about the [API changes automatic comment](https://github.com/OpenRailAssociation/osrd/pull/10200#issuecomment-3153925166)?

You already did by opening the `osrd-data` MR. AFAIK, it's the only consumer project of editoast's API that we know of ;)

> ETCS simulations are missing in this PR, should it be part of it?

For now, ETCS is not supported in STDCM - but I don't see what ETCS (with fixed blocks) would have to do with this?

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt:
```diff
@@ -486,3 +506,36 @@ private fun makeHeuristicsForPathfindingEdges(
+
+    // If there is no signal in the block after the operational point (waypointBlock), we consider
+    // the end of the block as the new stop position
+    val nextSignalPosition =
+        findNextSignalPositionFromBlockOffset(infra, waypointBlock.edge, waypointBlock.offset)
```
I don't think we need to bother with that. Only block-delimiting signals are relevant here (no point in stopping at a BAPR announcer signal). And block delimiting signals, well, they delimit blocks, can't move past one without changing block.

louisgreiner replied:
Oh yes, I did not thought of this case. Should we then rename the feature (flag and etc.) to precise that we specificly stop at block-delimiting signals only? Like "stops_at_end_of_block" or something similar?

louisgreiner replied:
Ok, done

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt:
```diff
@@ -486,3 +506,36 @@ private fun makeHeuristicsForPathfindingEdges(
+    val maxHeadOffset =
+        nextSignalPosition?.distance ?: infra.blockInfra.getBlockLength(waypointBlock.edge).distance
+
+    return PathfindingEdgeLocationId(waypointBlock.edge, Offset(min(maxTailOffset, maxHeadOffset)))
```
We can keep offset types as `Offset<Block>` without too much troubles. You'd need the specific min method here (`Offset.min`), and remove a few `.distance` here and there. (note: `Offset + Distance` is OK and returns an `Offset`)

louisgreiner replied:
Nice, done

eckter commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlockRequest.kt:
```diff
@@ -17,6 +17,7 @@ class PathfindingBlockRequest(
     val rollingStockSupportedSignalingSystems: List<String>,
     @Json(name = "rolling_stock_maximum_speed") val rollingStockMaximumSpeed: Double,
     @Json(name = "rolling_stock_length") val rollingStockLength: Double,
+    @Json(name = "stops_at_next_signal") val stopsAtNextSignal: Boolean,
```
I think I would have preferred if the flag were optional (with default to false), like how it's optional in the editoast API. The behavior with flag enable not that intuitive, while the behavior with no flag is what you'd normally expect. And the usecase is fairly niche. 

Ideally someone could start using the endpoint without having to wonder what that actually does and which value to chose.

louisgreiner replied:
I agree, done

eckter reviewed (approved):
Some small comments, but for the most part, LGTM

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt:
```diff
@@ -104,10 +103,29 @@ class PathfindingBlocksEndpoint(private val infraManager: InfraProvider) : Take
+                if (
+                    request.stopsAtEndOfBlock == true &&
+                        stepIndex != 0 &&
+                        stepIndex != request.pathItems.size - 1
```
We could end with a penultimate stop being pushed after the ultimate one, no?

What we could do about it:
* Add a test on a penultimate stop being before the ultimate one and in the same block, ultimate stop being before the end of the block.
* Cap the offset to the initial length of the TravelledPath (if we have it) while finding new stop position, or after when filling `waypoints` (to be applied for STDCM too in that case). It may require passing the positions of the possible last steps to compare if on the same edge?
* I don't think it's necessary to prevent the move before first stop, but if it's easy and clear we can add it.

ppitou replied:
@bougue-pe I handled this case and added tests if you could review it again.

bougue-pe commented on front/tests/assets/stdcm/linked-train/anterior-linked-train-table.json:
```diff
@@ -13,7 +13,7 @@
     "operationalPoint": "North_station",
     "code": "BV",
-    "endStop": "13:18",
+    "endStop": "13:17",
```
Just curious: why those 1 min changes?

louisgreiner replied:
It may be because the previous stop is now closer to this stop? I don't really know

bougue-pe replied:
It's now gone, resolving it.

It may have been that the application of safe approach speed around signal matches better the train's braking to a stop 🤷

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -37,6 +39,7 @@ fun getPathfindingBlockRequest(
         rs.maxSpeed,
         rs.length,
         null,
+        false,
```
nit: you may add a new param defaulting to `false`

also nit: you may use this method in your tests (there is a `VERY_SHORT_FAST_TRAIN`, `REALISTIC_FAST_TRAIN` and `VERY_LONG_FAST_TRAIN`)


Please feel very free to dismiss this as the fatigue on this much delayed PR is perfectly understandable.

louisgreiner replied:
Done

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -620,3 +623,255 @@ class PathfindingTest : ApiTest() {
+        )
+    }
+
+    private fun pathfindingRequest(
```
It looks like it can now be replaced by `checkPathfindingSuccess()`
The expected length will have to be provided, which is easy to have (may be manually compared to the case of `stopsAtEndOfBlock = false` if not already done for the last stop).

louisgreiner replied:
Done, the long train tests now fail. It outputs something interesting: 
```sh
expected: <[0m, 112050m, 26537.500m, 45900m]> but was: <[0m, 12320m, 26537.500m, 45900m]>
```

The first stop is way further than the next ones (because now the rs length is 100 000m)

It must be fixed

bougue-pe replied:
Oh, maybe add a check on the sort of offsets would be nice on the way, then?

I let you see if you want to handle it or not, resolving without it is fine by me.

ppitou replied:
Done

bougue-pe commented on python/osrd_schemas/osrd_schemas/train_schedule.py:
```diff
@@ -211,6 +213,11 @@ class TrainScheduleOptions(BaseModel):
         default=True,
         description="If false, the speed limits of the infrastructure are ignored in the standalone simulation",
     )
+    stops_at_end_of_block: bool = Field(
```
Should we change (bump from 0.8.19 to 0.8.20) the version of the Python package in [`osrd_schemas::pyproject.toml`](https://github.com/OpenRailAssociation/osrd/tree/dev/python/osrd_schemas/pyproject.toml#L3) ?
I guess it depends mostly on how we want to handle it in osrd-data.

louisgreiner replied:
Done

bougue-pe replied:
It (expectedly) got lost in rebases.

ppitou replied:
Done :P

bougue-pe reviewed (commented):
At least testing the case of penultimate "overtaking" the ultimate stop seems important.

I can give a hand around this, NP.

shenriotpro commented on python/osrd_schemas/osrd_schemas/train_schedule.py:
```diff
@@ -201,6 +201,8 @@ class TrainScheduleOptions(BaseModel):
     - `use_electrical_profiles` : use the electrical profiles for the standalone simulation
     - `use_speed_limits_for_simulation` : use the speed limits coming from the infrastructure for
        the standalone simulation
+    - `stops_at_end_of_block` : stops the train at the next block-delimiting signal instead of on the waypoint
```
nit: use imperative form (ditto for all occurrences)

ppitou replied:
Done

shenriotpro commented on python/osrd_schemas/osrd_schemas/train_schedule.py:
```diff
@@ -211,6 +213,11 @@ class TrainScheduleOptions(BaseModel):
     )
+    stops_at_end_of_block: bool = Field(
+        default=False,
+        description="If true, the train will stop at a new stop position: near next block-delimiting signal, \
```
nit: not sure new is the right word, it won't be new for long..

Khoyo replied:
"If true, stop positions will be shifted toward the next block-delimiting signal" ?

bougue-pe replied:
> "If true, stop positions will be shifted toward the next block-delimiting signal" ?

👍 that (and keeping the details of the second line below) would be nice IMO.

ppitou replied:
Done

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+        // The first intermediate stop is moved to the next block-delimiting signal, but the
+        // distance available is less than the length of the train, so it is moved to the end of the
+        // block
+        // (270 meters after the first intermediate stop)
```
nit: you got bitten by spotless (ditto on all the comments of the file) 😿
And my other suggestions might be affected too, of course 😓
```suggestion
        // block (270 meters after the first intermediate stop)
```

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+        // distance available is less than the length of the train, so it is moved to the end of the
+        // block,
+        // but the end of the block is also the destination so the intermediate stop remains the
+        // same
```
Not testing versus the end of the block here, just the destination:
```suggestion
        // distance available is more than the length of the train, so it would be moved by the length of the train (400m)
        // but this is also exactly the destination so the intermediate stop remains the same
```

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+    }
+
+    @Test
+    fun penultimateStopInsideDestinationBlock() {
```
To me this test is currently a corner case (exact limit) of the next test, and the naming may be misleading.
```suggestion
    fun penultimateStopNotMovedExactlyToDestination() {
```

bougue-pe replied:
This may be worth testing that it's correctly moved in the same block, with intermediateWaypoint at 12550 (previous signal is at 12500): i.e. not hitting destination.

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+        // The intermediate stop is moved to the next block-delimiting signal, but the
+        // distance available is less than the length of the train, so it is moved to the end of the
+        // block,
+        // but the end of the block after the destination so the intermediate stop remains the same
```
```suggestion
        // distance available is more than the length of the train, so it would be moved by the length of the train,
        // but that is after the destination so the intermediate stop remains the same
```

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+    }
+
+    @Test
+    fun penultimateStopCloseToDestination() {
```
```suggestion
    fun penultimateStopTooCloseToDestination() {
```

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt:
```diff
@@ -101,10 +102,33 @@ class PathfindingBlocksEndpoint(private val infraManager: InfraProvider) : Take
+                ) {
+                    val destinationTrack = request.pathItems.last().last()
+                    val destinationBlock =
+                        findWaypointBlocks(infra, destinationTrack, direction).last()
```
There's a problem here: the pathItems (destination also) can be OP (with multiple OP-parts, on different tracks) and each can be in different blocks (one for each signaling system protecting the range).
So we're only considering the last OP-part of the destination, and the last signaling-system, which may not be the ones actually used as destination.

It looks like:
* we should get all the blocks+offsets of the destination (function without `direction` param)
* then only in the `findStopPositionAtEndOfBlockConsideringRollingStock`, filter on the actual block of the waypoint, keeping only the minimal offset if multiple parts are in the block.

ppitou replied:
Thank you ! Done

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/stdcm/STDCMEndpoint.kt:
```diff
@@ -325,20 +327,34 @@ fun parseSteps(
+        .mapIndexed { index, it ->
+            STDCMStep(
+                if (index != 0 && index != pathItems.size - 1) {
+                    val destinationBlock = findWaypointBlocks(infra, it.locations).last()
```
Kind of the same problem here: we're considering (only) the last block-location of the current item (not the destination).

ppitou replied:
Done

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,265 @@ class PathfindingTest : ApiTest() {
+            expectedIntermediatePathItemPosition = listOf(Offset(intermediateStopDistance)),
+        )
+    }
+}
```
After the review, should we test also:
* 2 penultimate stops being too close from destination
* destination and penultimate stops having different parts and different signaling systems

?

IMO: "No, it's not worth the hastle" is a valid answer (especially for the second one), but if you think of an easy way to manage it...

bougue-pe replied:
✅ OK to resolve IMO (I let it open just in case it acts as a reminder for you)

✔️ Done for the first suggestion of two intermediate stops too close to destination.
✖️ Not done for the multiple parts and multiple signaling systems one (much heavier probably)

ppitou replied:
Thank you ! :dragon:

bougue-pe reviewed (commented):
The last 3 comments are especially important.

bougue-pe commented on core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingTest.kt:
```diff
@@ -623,3 +626,318 @@ class PathfindingTest : ApiTest() {
+
+        // The intermediate stop is moved to the next block-delimiting signal, but the
+        // distance available is more than the length of the train, so it would be moved by the
+        // length of the train, but that is after the destination so the intermediate stop remains the same
```
```suggestion
        // The intermediate stops are moved to the next block-delimiting signal, but the
        // distance available is more than the length of the train, so it would be moved by the
        // length of the train, but that is after the destination so the intermediate stops remain the same
```

ppitou replied:
Done

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/stdcm/STDCMEndpoint.kt:
```diff
@@ -417,7 +433,7 @@ private fun parseSimulationScheduleItems(
 }
 
-private fun findWaypointBlocks(
+fun findWaypointBlocks(
```
nit: better grouping it next to the other one in `PathfindingBlocksEndpoint.kt` (and if a rename is necessary, I'd go for "find`Directed`WaypointBlocks" for the other function).

ppitou replied:
Done

bougue-pe commented on core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt:
```diff
@@ -101,10 +104,32 @@ class PathfindingBlocksEndpoint(private val infraManager: InfraProvider) : Take
+                        stepIndex != request.pathItems.size - 1
+                ) {
+                    val destinationTrack = request.pathItems.last()
+                    val destinationBlock = findWaypointBlocks(infra, destinationTrack)
```
nit: this can be moved out of the `forEachIndexed` loop and computed only once.

ppitou replied:
Done

bougue-pe reviewed (commented):
✅ LGTM: The code logic matches the business logic as I understand it, thanks for the work!

I suspect tests are breaking now with the correction :-/
I can probably peer on fixing tests if needed.

bougue-pe reviewed (approved):
LGTM, thanks very much for the work!

There are 2 comments hanging (conveniently 😬 hidden by github's UX):
* https://github.com/OpenRailAssociation/osrd/pull/10200#discussion_r2569079043
* https://github.com/OpenRailAssociation/osrd/pull/10200#discussion_r2272878629

Also gl for the squash and the fix of e2e tests 💪

bougue-pe commented:
It looks like we are only missing a front review (+squash) before merge, no?

[What is the strategy for client repository on this](https://github.com/OpenRailAssociation/osrd/pull/10200#issuecomment-3153925166)? Merging this one in OSRD, and later activate it in client project (data import), I guess?

Khoyo commented:
Indeed, since this is a non-breaking change, we can merge it now and the downstream repos can use it at their leisure. For osrd-data, that'll be any point after this hits the deployed dev env.

SharglutDev commented on front/src/common/api/generatedEditoastApi.ts:
```diff
@@ -3933,6 +3936,7 @@ export type Margins = {
   values: string[];
 };
 export type TrainScheduleOptions = {
+  stops_at_end_of_block?: boolean;
```
Maybe we could also handle the case in `selectTrainToEdit` function (l.82) in `front/src/reducers/osrdconf/operationalStudiesConf/index.ts`

ppitou replied:
Done

SharglutDev reviewed (approved):
Lgtm (not tested), good job ! You can close the comment after fixing it.

bougue-pe commented on editoast/openapi.yaml:
```diff
@@ -14209,6 +14216,8 @@ components:
     TrainScheduleOptions:
       type: object
       properties:
+        stops_at_end_of_block:
```
If I understand correctly, this is optional for pathfinding (and TrainSchedule when storing the information).
And this is now automatic for STDCM, right?

🙏 Can you please edit the main (top) comment of the PR to express that (or correct me if I'm mistaken)?
Not sure if we can document it clearly in STDCM's API, also.

ppitou replied:
Is it clearer for the top comment ?

bougue-pe replied:
It is already clearer, yes, thanks.

---

For the trains simulated in the stdcm search environment, it depends on what was picked when importing data (like BASIC imports), no? (in practice we are gonna use it, but it's not tied to OSRD, just how we use it)

And I also plan to mention that:
* the feature can't be deactivated (front or back) for STDCM's path output.
* the feature can be switched on/off for pathfinding and train-schedule (operational studies) in the backend, but it can't be activated in the frontend.

If you don't correct me, I'm OK to change a bit more myself.

ppitou replied:
Yes it's good for me !

bougue-pe replied:
Thanks, done.

Khoyo reviewed (approved):
<img width="350" height="457" alt="Ship it" src="https://github.com/user-attachments/assets/687e888e-02e7-47e4-8b4d-e22d7badab31" />
